{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# for questions, contact karenmei@ucsd.edu\n",
    "\n",
    "#Goal: To evaluate the custom hierarchy (Metric #2)\n",
    "\n",
    "#Metric 2: To create a pipeline for testing for how well the custom ontology maps onto reference ontology using DDOT \n",
    "#How well does the model capture the known structure of the reference ontology? (evaluated by alignment to the Gene Ontology, i.e. how many GO terms significantly overlap gene modules in your model?)\n",
    "\n",
    "#\t   code for customizing ontologies: DDOT: https://github.com/michaelkyu/ddot/blob/master/examples/Tutorial.ipynb\n",
    "\n",
    "import sys\n",
    "sys.path.append('C:\\\\Users\\\\Anubhav\\\\Documents\\\\GitHub\\\\ddot')\n",
    "\n",
    "import pandas as pd\n",
    "import networkx as nx\n",
    "import numpy as np\n",
    "import os\n",
    "import ddot\n",
    "from ddot import Ontology\n",
    "import csv\n",
    "import scipy.stats as ss\n",
    "from scipy.stats import hypergeom\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "os.environ['KMP_DUPLICATE_LIB_OK']='True'\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#Create dictionary of all the term_names and description (in this example: GO)\n",
    "#i.e. example in this dictionary: key is 'GO:0005694'; value: 'chromosome'\n",
    "#input: Term is a GO term; for example: 'GO:0005694'\n",
    "\n",
    "def Find_GO_Term_Desc(Term):\n",
    "\twith open('goID_2_name.tab', 'r') as f:\n",
    "\t\tGO_terms=[r for r in csv.reader(f, delimiter='\\t')]\n",
    "\tterm=[]\n",
    "\tdesc=[]\n",
    "\tfor i in range(len(GO_terms)):\n",
    "\t\tterm.append(GO_terms[i][0])\n",
    "\t\tdesc.append(GO_terms[i][1])\n",
    "\tdictionary=dict(list(zip(term, desc)))\n",
    "\tdescription=dictionary[Term]\n",
    "\treturn description\n",
    "\n",
    "\n",
    "#Generate custom ontology file\n",
    "def Generate_Ontology_File(Term, file_name):\n",
    "\tndex_server ='http://public.ndexbio.org' \n",
    "\tgo_human = Ontology.from_ndex('http://public.ndexbio.org/v2/network/16565851-4bfc-11e9-9f06-0ac135e8bacf')\n",
    "\tont= go_human.focus(Term)\n",
    "\tont.to_table(file_name)\n",
    "\treturn ont\n",
    "\n",
    "#Convert the custom ontology into the desired descriptions and gene names (in this example: GO descriptions and gene IDs)\n",
    "#the input to this function is ont, which is the output of Generate_Ontology_File\n",
    "def Find_GO_Focus_GeneDict(ont):\n",
    "\tont = ont.propagate(direction='forward', gene_term=True, term_term=False)\n",
    "\tterms_to_genes=ont.term_2_gene\n",
    "\tGO_Desc=[]\n",
    "\tgene_names=[]\n",
    "\tfor key in terms_to_genes:\n",
    "\t\tterm_name=key\n",
    "\t\tGO_Desc.append(term_name)\n",
    "\t\tgenes_in_terms=terms_to_genes[key]\n",
    "\t\tgenes_list=[]\n",
    "\t\tfor item in genes_in_terms:\n",
    "\t\t\tgene_name=ont.genes[item]\n",
    "\t\t\tgenes_list.append(gene_name)\n",
    "\t\tgene_names.append(genes_list)\n",
    "\tGO_term_gene=list(zip(GO_Desc, gene_names))\n",
    "\tGO_dict=dict(GO_term_gene)\n",
    "\treturn GO_dict\n",
    "\n",
    "    \n",
    "\n",
    "#Find the enriched terms within the custom ontology\n",
    "#function input: ont (output of Generate_Ontology_File)\n",
    "#               syn_ont (output of Find_GO_Focus_GeneDict)\n",
    "#               test_gene_list (list of genes to be tested)\n",
    "def Find_Enrichment(ont, syn_ont, test_gene_list):\n",
    "\tnum_ont_genes=len(ont.genes)\n",
    "\tnum_intersect_test_ont=len(set(ont.genes)&set(test_gene_list))\n",
    "\toverlap=[]\n",
    "\tfor key in syn_ont:\n",
    "\t\tpair=[]\n",
    "\t\tont_gene_list=syn_ont[key]\n",
    "\t\tsize_ont_genes=len(ont_gene_list)\n",
    "\t\toverlap_num=len(list(set(ont_gene_list)&set(test_gene_list)))\n",
    "\t\tpair.append(key)\n",
    "\t\tpair.append(overlap_num)\n",
    "\t\tpair.append(size_ont_genes)\n",
    "\t\toverlap.append(pair)\n",
    "\tp_values=[]\n",
    "\tfor item in overlap:\n",
    "\t\tintersect_termgenes_testgenes=item[1]\n",
    "\t\tnum_term_genes=item[2]\n",
    "\t\tp=hypergeom.sf(intersect_termgenes_testgenes-1, num_ont_genes,num_intersect_test_ont, num_term_genes)\n",
    "\t\tp_values.append(p)\n",
    "\tp_adj=multipletests(p_values,method='fdr_bh')\n",
    "\tboolean_p=p_adj[0]\n",
    "\tidx_true=[i for i, e in enumerate(boolean_p) if e != False]\n",
    "\tp_adj=p_adj[1]\n",
    "\tmin_p=min(p_adj)\n",
    "\tidx_min_p=[i for i, e in enumerate(p_adj) if e == min_p]\n",
    "\tmost_enriched_terms=[]\n",
    "\tfor item in idx_min_p:\n",
    "\t\tterm_padj_pair=[]\n",
    "\t\tenriched=overlap[item][0]\n",
    "\t\tGO_term_padj=p_adj[item]\n",
    "\t\tterm_padj_pair.append(enriched)\n",
    "\t\tterm_padj_pair.append(GO_term_padj)\n",
    "\t\tmost_enriched_terms.append(term_padj_pair)\n",
    "\ttrue_terms=[]\n",
    "\tfor item in idx_true:\n",
    "\t\ttrue_pair=[]\n",
    "\t\ttrue_term=overlap[item][0]\n",
    "\t\ttrue_padj=p_adj[item]\n",
    "\t\ttrue_pair.append(true_term)\n",
    "\t\ttrue_pair.append(true_padj)\n",
    "\t\ttrue_terms.append(true_pair)\n",
    "\t#print ('terms above threshold:', true_terms)\n",
    "\treturn true_terms\n",
    "\n",
    "#Find the number of enriched modules\n",
    "#function input: ont1, ont2\n",
    "#ont1 (ontology 1); ont2 (ontology 2)\n",
    "def Num_Enriched_Modules(ont1, ont2):\n",
    "\tchr_ont=Find_GO_Focus_GeneDict(ont1)\n",
    "\t#print (chr_ont)\n",
    "\ttest_ont=Find_GO_Focus_GeneDict(ont2)\n",
    "\t#print (test_ont)\n",
    "\tenriched_modules=[]\n",
    "\tfor key in test_ont:\n",
    "\t\ttest_gene_list=test_ont[key]\n",
    "\t\tif len(test_gene_list)!=0:\n",
    "\t\t\tenriched_terms=Find_Enrichment(ont1, chr_ont, test_gene_list)\n",
    "\t\t\tenriched_modules.append(enriched_terms)\n",
    "\t\telse:\n",
    "\t\t\tcontinue\n",
    "\tenriched_modules = [x for x in enriched_modules if x != []]\n",
    "\tenriched_modules = [item for sublist in enriched_modules for item in sublist]\n",
    "\tenriched_term_names=[]\n",
    "\tfor item in enriched_modules:\n",
    "\t\tenriched_term_name=item[0]\n",
    "\t\tenriched_term_names.append(enriched_term_name)\n",
    "\tunique_enriched_term_names=set(enriched_term_names)\n",
    "\t#print (unique_enriched_term_names)\n",
    "\tprint ('Num of Enriched Modules', len(unique_enriched_term_names))\n",
    "\treturn "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "num synapse ontology terms 193\n",
      "num custom ontology terms 110\n",
      "Num of Enriched Modules 59\n"
     ]
    }
   ],
   "source": [
    "#Generate custom ontology 1 from synapse branch from human GO\n",
    "#ontology_file=Generate_Ontology_File('GO:0045202', 'ont1.txt')\n",
    "ont1=Ontology.from_table('synapse.txt')\n",
    "print ('num synapse ontology terms', len(ont1.terms))\n",
    "\n",
    "ont2=Ontology.from_table('finalontology3.txt')\n",
    "print ('num custom ontology terms', len(ont2.terms))\n",
    "\n",
    "Num_Enriched_Modules(ont1, ont2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1177 genes, 193 terms, 2378 gene-term relations, 315 term-term relations\n",
      "node_attributes: []\n",
      "edge_attributes: []\n",
      "788 genes, 124 terms, 2541 gene-term relations, 126 term-term relations\n",
      "node_attributes: []\n",
      "edge_attributes: []\n"
     ]
    }
   ],
   "source": [
    "print(ont1)\n",
    "print(ont2)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
